### PR TITLE
ConfigV2: Mixins

### PR DESCRIPTION
**Pull request by jkellerer**
Provides configuration `mixins` (multiple inheritance). See also "templates" in https://github.com/creativeprojects/resticprofile/issues/80#issuecomment-1140060881

Status
- [x] Implementation
- [x] Tests
- [x] Documentation
- [x] Review of details (e.g. parameter append syntax, variable syntax, default-vars)
- [ ] Code review

Example:
```yaml
version: 2

mixins:
  healthchecks:
    default-vars:
        URL: https://hc-ping/uuid

    send-before:
        method: HEAD
        url: ${URL}/start

    send-after:
        method: HEAD
        url: ${URL}

    send-after-fail:
        method: POST
        url: ${URL}/fail
        body: "${ERROR}\n\n${ERROR_STDERR}"
        headers:
          - name: Content-Type
            value: "text/plain; charset=UTF-8"

profiles:
  # simple syntax
  with_monitoring:
    use:
      - healthchecks
    backup: 
      source: /

  # full syntax
  with_monitoring_and_uuid:
    use:
      - name: healthchecks
        vars:
          URL:  https://hc-ping/another-uuid
    backup: 
      source: /
```

List append example ("key..." appends to a list, "...key" prepends to a list)
```yaml
version: 2

mixins:
  vm-snapshot:
    default-vars:
        LIVEDATA: /var/db
    
    ...run-before:
      - virsh snapshot-create-as --domain "${VM}" --name "${VM}-backup" --diskspec "vda,file=${LIVEDATA}/${VM}-livedata.qcow2" --disk-only --atomic --no-metadata --quiesce
      
    run-finally...:
      - [ -e "${LIVEDATA}/${VM}-livedata.qcow2" ] && virsh blockcommit --domain "${VM}" vda --wait --active && virsh blockjob --domain "${VM}" "${LIVEDATA}/${VM}-livedata.qcow2" --pivot && rm -f "${LIVEDATA}/${VM}-livedata.qcow2"

profiles:
  vms:
    backup:
      use:
        - {name: vm-snapshot, VM: my-app-server}
        - {name: vm-snapshot, VM: my-docker-server}

      source:
        - /opt/vms/my-app-server*
        - /opt/vms/my-docker-server*

      run-before:
        - echo all snapshots created
      run-after:
        - echo all vm snapshots backed-up
      run-finally:
        - echo removing all existing snapshots now
```

### CODE REVIEW

**Comment by jkellerer** [comment]
Hi @creativeprojects , would you mind reviewing the current implementation and concept whether this meets your expectations.

Once you are fine, I'll check how I can add the changes to the documentation and after that I'll mark the PR as ready.

**Comment by creativeprojects** [comment]
It looks good to me so far 👍🏻 

I was thinking about using `<` and `>` instead of `+` but it's not allowed on TOML so far. I haven't checked but it's probably not allowed on the other formats either.

Would that make it more readable to put it in the values instead?

like ["++", "other value"] or ["other value", "++"]? Just thinking out loud here.

**Comment by creativeprojects** [comment]
If the default tags are messing up with some of the user's values, we could make both tags configurable in the global section though?

**Comment by jkellerer** [comment]
### Suffixes for appending

While I think in general it is a good idea to make things configurable, for the suffixes (or tags) we probably should decide on a syntax to simplify documentation and sharing of snippets.

Putting append or prepend logic in the value is an option but I think it would confuse me a bit in a config file, might just be me though.

TOML officially allows the following chars `[A-Za-z0-9_-]` in config keys (unless they are quoted).

Only valid in YAML, TOML requires quoted key for the following variations:

```plain
property>        run-after-fail>: ...
property<        run-after-fail<: ...

property>>       run-after-fail>>: ...
property<<       run-after-fail<<: ...

property+>       run-after-fail+>: ...
property->       run-after-fail->: ...

property++       run-after-fail++: ...
property+0       run-after-fail+0: ...

property__++     run-after-fail__++: ...
property__0+     run-after-fail__0+: ...

property+LAST    run-after-fail+LAST: ...
property+FIRST   run-after-fail+FIRST: ...
```

Valid for any config file format:

```plain
property__LAST         run-after-fail__LAST: ...
property__FIRST        run-after-fail__FIRST: ...

property__APPEND       run-after-fail__APPEND: ...
property__PREPEND      run-after-fail__PREPEND: ...

APPEND__property     APPEND__run-after-fail: ...
PREPEND__property    PREPEND__run-after-fail: ...
```

I would tend to `property__APPEND` / `property__PREPEND` and maybe allow an alternate short syntax from the list above (e.g. `property__++` / `property__0+`). There should be no overlaps with existing config options and also this is only allowed within templates, not within other areas of the configuration.

### Naming it

At several places we already use "template" and mean go templates for something (mostly to generate markup of some sort). It probably may confuse users when this feature is also called "templates". I've used the term "inline-template" as a working title to have something for coding but it is probably not too good.

Logically this feature is more related to `mixins` (multiple inheritance) and/or config `snippets`, how about:

```yaml
version: 2

mixins:
  exclude-hidden:
    backup:
      exclude__APPEND: [".*", "*~"]

profiles:
  projects:
    use: ["exclude-hidden"]
    backup:
      source: /projects/src
```

**Comment by creativeprojects** [comment]
Thanks for your research, this is very much appreciated 🙇🏻 

I do like the `PREPEND` and `APPEND`. This is easy to remember, and easy to read. I can't really decide before or after the field name, maybe both? it's just a case of `HasPrefix || HasSuffix` after all.

I guess it depends how you're going to read it in your head:
- append to this field using these values
- take this field and append these values

I thought about another cool one, which works in YAML (but sadly not TOML):
`...for_prepend` and `for_append...`

I guess we could allow it on YAML and only the APPEND/PREPEND on TOML? maybe

---

Now your idea of `mixins` and `use` is spot on!
I also had concerns about providing two different ways of doing templating, and that closes the case very nicely 😄

**Comment by jkellerer** [comment]
Successfully renamed to `mixins` and changed syntax to `__APPEND` (or `key...`) and `__PREPEND` (or `...key`). Even the following toml works fine:
```toml
[mixins.exclude-macos]
"exclude..." = [
    ".DS_Store",
]
```

You might notice that I made key-delimiter as used by `viper` configurable at locations where it wasn't. It is "." and should conflict with the short syntax but surprisingly, it works without changing it. Should it break with an update, then it can be changed and tests will still succeed.

Next is documentation. Think I'll start with a new page for "inheritance" that explains "inherit", "mixin" and how flags are passed from top to bottom as I this topic isn't really covered so far on its own?

**Comment by creativeprojects** [comment]
I'm amazed the 3 dots work with TOML, that's excellent news.

For the documentation, I'm afraid there's nothing for inherit. I started this project on my own requirements and since there were plenty similar ones I didn't bother much with documentation. Inherit was implemented from the beginning.

So I was just relying on the user reading a big configuration file to understand it all 😄 :
https://creativeprojects.github.io/resticprofile/configuration/examples/

**Comment by jkellerer** [comment]
@creativeprojects , PR is awaiting a review :)

**Comment by creativeprojects** [comment]
Cool, I'll have a look 👍🏻 

I approved #113 but I don't know if you want to wait for this one

**Comment by jkellerer** [comment]
The 2 PRs should not have conflicts and can be merged in any order (though the documentation enhancements in this PR assume that #113 is already merged).

**Comment by creativeprojects** [comment]
Right, I'll merge #113 now then 😉

**Comment by jkellerer** [comment]
Looks like it is still valid, will rebase and run tests just to be sure 😀

File: docs/content/configuration/inheritance/_index.md
Comment (creativeprojects) [inline docs/content/configuration/inheritance/_index.md:13]: you can use the macro {{< ref "/configuration/hooks" >}} to generate a url. The good thing is that it's going to fail when we rename a chapter 😉

File: docs/content/configuration/inheritance/_index.md
Comment (creativeprojects) [inline docs/content/configuration/inheritance/_index.md:15]: typo: configuration

File: docs/content/configuration/include/_index.md
Comment (creativeprojects) [inline docs/content/configuration/include/_index.md:102]: as such it's not part of the "merging" process, but should we also note where the go template gets resolved in the grand scheme of things?

**Review by creativeprojects — COMMENTED** [review]
It all looks good to me 😄 

I tried a few mixins and they all got resolved as expected.

I just have a small comment: should we hide the "use" section in the `show` command? since the profile has been merged at that point.

Nice work 👍🏻

File: docs/content/configuration/inheritance/_index.md
Comment (jkellerer) [reply docs/content/configuration/inheritance/_index.md:13]: Was using IDE auto-complete but yeah, compile time check makes sense 😀

File: docs/content/configuration/include/_index.md
Comment (jkellerer) [reply docs/content/configuration/include/_index.md:102]: Makes perfect sense to me. Added it as introduction to "loading" as it's important that this is understood.

**Comment by jkellerer** [comment]
Regarding hiding `use`: Had first set the use property to nil to hide it by removing it (`viper` api allows setting but not removing properties). But viper has very strange behaviour when setting `some.path.property`. This sets `some` and `some.path` to an empty map so setting `profiles.profile.use` to `nil` effectively also sets `profile` to undefined.

It could be removed in `show` and this would be consistent with `inherit`. Hmm, probably better to be consistent.

**Review by creativeprojects — APPROVED** [review]
Great work!

Now, if I remember well, the only thing missing for v2 is the scheduling (since I decided to move it to its own section). And then I need to go round everything to make sure I haven't forgotten some bits...

Thanks 😉

**Pull request merged**
